[4.2.x] Remove frontend usage of federation, updates query to only use user defaults if flag is passed

andrewkfiedler opened the pull request:
- Updates query to by default not use the user specified settings.
 - Removes code paths that are no longer needed.

Testing:  
 - Update your user settings to have a selected set of sources (such as slow).
 - Verify creating a new search will utilize those sources.
 - Verify that when visiting the id page of a metacard, the query that goes out does not use the users setting for sources (set it to something that the metacard isn't in, or just look at the network).

andrewkfiedler committed:
Remove frontend usage of federation, updates query to only use user defaults if flag is passed

 - Updates query to by default not use the user specified settings.
 - Removes code paths that are no longer needed.

andrewkfiedler committed:
Starts a typed version of query

 - We can slowly move over to using a more typed out version of query.

andrewkfiedler commented:
I just realized I need to do a similar thing for the ephemeral filtering and sorting (they are currently applied in scenarios they should not be, ex: add a filter after searching, then go to an id page you should be able to see and it will be inadvertently filtered).  I should be able to add a similar fix to the sources fix.

andrewkfiedler committed:
Fix issue with orphaned queries sticking around

 - Fixes issue with orphaned queries sticking around.  Basically, since certain listeners (for filtering / sorting, result count), lived in the query model itself, queries that were no longer being used by views would stil
l react / run.  Now, the listeners are tied directly to the view lifecycle with useQuery / useUserQuery.
 - Adds useListener as a convenience method for the most common backbone usage within react.
 - Adds TypedQuery as a start to wrapping around the untyped version.  It has some hooks, and the typical options that we expose for queries.  The transformDefaults option allows altering defaults (namely injecting user defaults if desired), transformFilterTree gets called right before we query (mix in user ephemeral filters, or anything else you want), transformSorts is similar.  There is also the familiar limitToDeleted, limitToHistoric, but I think we can remove these in favor of just using transformFilterTree.
 - Removes a few dead code paths, like the old metacard route.

andrewkfiedler committed:
Fix formatting / typescript

jordanwilking commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/pages/metacard.tsx:
```diff
@@ -3,33 +3,57 @@ import { hot } from 'react-hot-loader'
+            })
+          }
+        })
+        .concat(
+          new FilterClass({
+            type: '=',
+            value: '-1',
+            property: 'id',
+          })
+        )
```
❓ So, if the file doesn't have any children, we default a filter with the file id and these lines. What does this do?

andrewkfiedler replied:
I'm trying to remember why exactly, but it's been so long I can't.  From what I recall the queries didn't finish correctly without this.

However, at this point it might be a left over of an old limitation that is no longer necessary.

jordanwilking reviewed (approved):
Liking the hooks 👍

jordanwilking commented:
ℹ️ Hero Successful ℹ️ 

#### Steps

1. Changed user settings to slow(offsite) in Settings -> Search Options
2. Navigated to search and verified that it had my search
    ![image](https://user-images.githubusercontent.com/22667297/114608458-91111800-9c52-11eb-8dd9-d8ccc57d0c02.png)
3. Ingested a file and went to its id page. Verified that the network didn't use the user settings
    ![image](https://user-images.githubusercontent.com/22667297/114608654-c289e380-9c52-11eb-9349-cc4ff785e5c3.png)
4. Removed all sources in settings and tried again to verify that the id query was unaffected.
5. Added a harvested source, set that as my source in settings, and tried again. The metacard Id route didn't use the user settings sources.
    ![image](https://user-images.githubusercontent.com/22667297/114609088-3f1cc200-9c53-11eb-95af-83ef8632c1ad.png)

andrewkfiedler committed:
Update to allow passing no arguments to useQuery, useUserQuery, fix DEFAULT_USER_QUERY_OPTIONS

 - original defaults were not being splatted, so original defaults would just be lost
 - updated the two hooks to allow passing 0 arguments

andrewkfiedler committed:
Backport of menustate with some additional helpers

samuelechu commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/pages/home.tsx:
```diff
@@ -206,24 +206,24 @@ const LeftBottom = ({ selectionInterface }: { selectionInterface: any }) => {
+  const location = useLocation()
+  let urlBasedQuery = location.search.split('?defaultQuery=')[1]
+  const [queryModel] = useUserQuery(decodeUrlIfValid(urlBasedQuery))
+  console.log(queryModel)
```
remove?

samuelechu commented on ui-frontend/packages/catalog-ui-search/src/main/webapp/component/query-settings/source-selector.tsx:
```diff
@@ -87,27 +87,13 @@ const shouldBeSelected = ({
  * If it includes 'remote', that that means everything remote.  All other values are singular selections of a source.
```
Does this type need to be updated to remove 'enterprise'? https://github.com/codice/ddf-ui/blob/7fbc7060ccceb0f98184eb59942a742a999c8bab/ui-frontend/packages/catalog-ui-search/src/main/webapp/component/query-settings/source-selector.tsx#L21

andrewkfiedler replied:
Sort of, it should be 'all' 'fast' 'slow' string now

andrewkfiedler committed:
Removes console log and updates source selector type
